packages/adapters-library/src/adapters/stargate/products/farm/stargateFarmAdapter.ts

```diff
@@ -0,0 +1,454 @@
+      ? LpStaking__factory.connect(
+          lpStakingAddress,
+          this.provider,
+        ).pendingStargate(...rewardFunctionInput)
+      : LpStakingTime__factory.connect(
+          lpStakingAddress,
+          this.provider,
+        ).pendingEmissionToken(...rewardFunctionInput))
+
+    return [
+      {
+        ...rewardToken,
+        type: TokenType.UnderlyingClaimable,
+        balanceRaw: pendingStargateReward,
+      },
+    ]
+  }
+
+  async getWithdrawals(input: GetEventsInput): Promise<MovementsByBlock[]> {
+    return await this.getMovements({
+      ...input,
+      filterType: 'withdrawal',
+    })
+  }
+
+  async getDeposits(input: GetEventsInput): Promise<MovementsByBlock[]> {
+    return await this.getMovements({
+      ...input,
+      filterType: 'deposit',
+    })
+  }
+
+  private async getMovements({
+    protocolTokenAddress,
+    fromBlock,
+    toBlock,
+    userAddress,
+    filterType,
+  }: GetEventsInput & {
+    filterType: 'deposit' | 'withdrawal'
+  }): Promise<MovementsByBlock[]> {
+    const protocolToken =
+      await this.getProtocolTokenByAddress(protocolTokenAddress)
+
+    const lpStakingContract =
+      protocolToken.lpStakingType === 'LpStaking'
+        ? LpStaking__factory.connect(
+            protocolToken.lpStakingAddress,
+            this.provider,
+          )
+        : LpStakingTime__factory.connect(
+            protocolToken.lpStakingAddress,
+            this.provider,
+          )
+
+    const filter =
+      filterType === 'deposit'
+        ? lpStakingContract.filters.Deposit(
+            userAddress,
+            protocolToken.poolIndex,
+            undefined,
+          )
+        : lpStakingContract.filters.Withdraw(
+            userAddress,
+            protocolToken.poolIndex,
+            undefined,
+          )
+
+    const events = await lpStakingContract.queryFilter(
+      filter,
+      fromBlock,
+      toBlock,
+    )
+
+    return events.map((event) => {
+      const { amount } = event.args!
+
+      return {
+        protocolToken: {
+          address: protocolToken.address,
+          name: protocolToken.name,
+          symbol: protocolToken.symbol,
+          decimals: protocolToken.decimals,
+        },
+        blockNumber: event.blockNumber,
+        transactionHash: event.transactionHash,
+        tokens: [
+          {
+            address: protocolToken.address,
+            name: protocolToken.name,
+            symbol: protocolToken.symbol,
+            decimals: protocolToken.decimals,
+            type: TokenType.Underlying,
+            blockNumber: event.blockNumber,
+            balanceRaw: amount,
+          },
+        ],
+      }
+    })
+  }
+
+  /**
+   * There are no specific events for rewards, which occur as simple transfers whenever a Deposit or Withdrawal is actioned
+   * For this reason, we need to fetch all Deposit and Withdrawal events and then filter out the reward transfers
+   */
```

Comment:
These are the type of scenarios, among others, are the reason why I dont want us to rule out having an indexer....hence why I want @xavier-brochard to at least explorer the indexer solution for ETH2